### ISSUE TITLE
Configuration for sensitive values not behaving properly for lifecycle { ignore_changes = {my_option} }

### ISSUE THREAD

**Issue by davidtc7** [auth0/terraform-provider-auth0#860]
### Checklist

- [X] I have looked into the [README](https://github.com/auth0/terraform-provider-auth0#readme) and have not found a suitable solution or answer.
- [X] I have looked into the [documentation](https://registry.terraform.io/providers/auth0/auth0/latest/docs) and have not found a suitable solution or answer.
- [X] I have searched the [issues](https://github.com/auth0/terraform-provider-auth0/issues) and have not found a suitable solution or answer.
- [X] I have upgraded to the [latest version](https://github.com/auth0/terraform-provider-auth0/releases/latest) of this provider and the issue still persists.
- [X] I have searched the [Auth0 Community](https://community.auth0.com/c/sdks/5) forums and have not found a suitable solution or answer.
- [X] I agree to the terms within the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md).

### Description

Current infra at our organization do not allow managing secret values through terraform due to it being present in plan and state. We generally manage secret outside of the terraform by doing following
1. put a dummy value in the terraform configuration.
2. put `lifecycle { ignore_changes = [ my_secret_configuration ]}` to let terraform ignore the values for that specific configuration.
3. update the actually secret value in management UI manually later.
4. next time we apply the terraform, terraform will not override the value with the dummy value in the terraform due to `ignore_changes`.

However, when I try to apply the same pattern for Auth0 resources, it does not work properly. At step 4, Auth0 provider will override the value with dummy value in the terraform even when it's not shown in the terraform plan.

I have observed the issue being present for following configurations:
1. `options.configuration` for custom DB connection.
2. `options.client_secret` for Google OAuth2 connection.
3. `secrets` for action scripts.
4. `sink.splunk_token` for Splunk log stream integration.
There may be more but these are the only ones I tested personally.

### Expectation

When `lifecycle {ignore_changes}` is applied to a field, Auth0 provider should ignore the values in the terraform during subsequent updates to the resource.

If such usage pattern is not allowed by the terraform provider, maybe the provider should give an explicit error instead of updating the value even if it's not showing in the plan.

### Reproduction

1. Apply following resource:
```
resource "auth0_action" "test_action" {
  name    = format("Test Action %s", timestamp())
  runtime = "node18"
  deploy  = true
  code    = <<-EOT
   exports.onExecutePostLogin = async (event, api) => {
     console.log(event.secrets.secret); 
   };
  EOT

  supported_triggers {
    id      = "post-login"
    version = "v3"
  }

  secrets {
    name  = "secret"
    value = "terraform_fake_value"
  }

  lifecycle {
    ignore_changes = [secrets]
  }
}
```
2. Update the secret value in Auth0 Management UI. Test the action, you should see the value you added being printed in output.
3. Make some small changes to the script and apply the terraform again.
4. Refresh the management UI page and test the action again, you should see `terraform_fake_value` being printed in the output.

### Auth0 Terraform Provider version

1.0.0

### Terraform version

1.5.7

**Comment by sergiught** [comment]
Hey @davidtc7 👋🏻 

Thanks for taking the time to raise this with us. The Terraform `ignore_changes` feature is intended to be used when a resource is created with references to data that may change in the future, but should not affect said resource after its creation. However this will not prevent updating the value if the config already contains certain config that needs to be applied, and this will not be shown in the diff because of the lifecycle `ignore_changes` block.

So in concrete terms, even if you have lifecycle `{ ignore_changes = [secrets] }` if you also have a block that defines the secrets as you have above `secrets { name = "secret" value = "terraform_fake_value" }` that will still get applied because it's present in the config. If you want the value to be ignored you need to remove the secrets block from the config. 

Example (notice how secrets is removed from the config and thus we won't end up overriding it any more):
```
resource "auth0_action" "test_action" {
  name    = format("Test Action %s", timestamp())
  runtime = "node18"
  deploy  = true
  code    = <<-EOT
   exports.onExecutePostLogin = async (event, api) => {
     console.log(event.secrets.secret); 
   };
  EOT

  supported_triggers {
    id      = "post-login"
    version = "v3"
  }

  lifecycle {
    ignore_changes = [secrets]
  }
}
```

Also `ignore_changes` is working as intended here -- it is not there as a mechanism to omit data from the state.

Further references: https://github.com/hashicorp/terraform/issues/516

So in summary, this is not a bug, but how the `ignore_changes` block is currently working and it's not just something that applies to sensitive values, the exact same behavior can be experienced with changes to the `code` block even if that is added to the `ignore_changes` list. Please consider opening an issue with https://github.com/hashicorp/terraform/issues in order to review this mechanism with the Terraform maintainers. 

Let me know if you have any questions, happy to clarify further.

**Comment by gojanpaolo** [comment]
I get a similar issue with `auth0_connection` `options` block.
```
resource "auth0_connection" "test" {
  name     = "test"
  strategy = "auth0"

  options {
    configuration = {
      foo = "bar"
    }
  }

  lifecycle {
    ignore_changes = [
      options[0].custom_scripts
      # more attributes ignored
    ]
  }
}
```

We'd like to managed `configuration` in terraform but most of the other attributes are managed outside.
If we apply a change to `configuration`, other `options` attributes (e.g. `custom_scripts`) get overwritten with the default values.

**Comment by sergiught** [comment]
Hey @gojanpaolo 👋🏻 

Unfortunately adding a lifecycle ignore changes block won't work for any of the connection options, as due to API limitations the inner options block behaves like a PUT instead of a PATCH, so all options configs need to be set in order to prevent erasing them. 

For enhanced security, store your secrets in a dedicated secret store. These are specialized databases designed to securely store sensitive data and tightly control access. Some popular options to consider include HashiCorp Vault, AWS Secrets Manager, AWS Parameter Store, and Google Cloud Secret Manager. Then have that hooked within the options secrets you want to manage so the auth0 terraform provider is made aware of them.

**Comment by davidtc7** [comment]
@sergiught Hey, the challenges are not how we store the secrets nor how we pass those secrets into terraform.

Terraform has these two things called terraform plan and terraform state (which I hope everyone is aware). We generate a terraform plan (here I mean an executable plan, not console output) as artifact of the CICD process and applying the plan in a separate CICD stage. And each time you apply, terraform also record all the configuration as terraform state. **_Both terraform plan, and terraform state contains your full configuration including all your secrets_**, it's far more challenging to secure these when your infrastructure is built around not having to do this. Also having secrets present in these many places just increase the attack surface a lot more.

**Comment by sergiught** [comment]
Hey @davidtc7 👋🏻 

I totally understand where you're coming from but I'd advise to follow the recommendations from Hashicorp around dealing with sensitive data within the state: https://developer.hashicorp.com/terraform/language/state/sensitive-data and take appropriate actions to update your infrastructure to be as secure as possible. It's inevitable that the terraform state will contain secrets.

**Comment by davidtc7** [comment]
@sergiught yes managing sensitive data through terraform is do able now by securing all those things, but it has not been considered a good practice just due to the fact that there are more ways your secrets can potentially be exposed. Currently we are building our own pipeline to secure the plan and states, but I just hope that when Auth0 build out the platform to also have considerations for those who do not wish to manage sensitive data through terraform.

**Comment by sergiught** [comment]
Hey folks, while I understand the ask here, it's something that we unfortunately cannot do within the current behavior of the v1 release. But it's something we can put on our radar for a v2 in the future. So unfortunately we'll have to close this down for the moment.
